### PR TITLE
seo

### PR DESCRIPTION
**Pull request by kiahjh**
At some point a while back the `/audiobooks` route got deleted, so I added it back in.

### CODE REVIEW

**Review by jaredh159 — CHANGES_REQUESTED** [review]
Coming along nicely. 👍

Regarding SEO, what i can see here looks good, but it appears that you did not address any of the dynamically generated pages, namely the _friend_ and _document_ pages. they have their own custom generated SEO titles, descriptions, and og images.

Regarding the restored audio page: maybe i missed this the first time, but there are a couple issues here. **a)** the featured audio book cards at the top have something amiss about their buttons, check the alignment vs prod; and **b)** in prod we sort all of the audiobooks by when they were added, most recent first. you you need to do the same and ensure the order correctly matches prod.

File: apps/next-evans/__tests__/seo.spec.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import { describe, it, expect } from 'vitest';
+import { documentPageDescription, friendPageDescription } from '../lib/seo';
+
+describe(`friendPageDescription()`, () => {
```
Comment (jaredh159) [inline apps/next-evans/__tests__/seo.spec.ts:4]: these tests are nice 👍

...but did you check on and convert my old tests for this? the file is called `seo.spec.ts`

speaking of which, maybe take a look at https://vitest.dev/api/#test-each and make sure you're clear on the `test.each` style of tests, i find i use it probably 50% or more of the time, it allows you to write more tests with less boilerplate, and perhaps most importantly, decreases my reluctance to add another test.

you don't necessarily have to delete these tests, but ideally we should keep the others at least as well, there may be hard-won edge cases covered there 👍

File: apps/next-evans/components/core/Seo.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import React from 'react';
+import { LANG } from '@/lib/env';
+
+type Props = {
+  title: string;
+  description: string;
+  withoutEnding?: boolean;
+} & (
+  | { documentPage: false }
+  | { documentPage: true; documentSlug: string; authorSlug: string; edition: string }
```
Comment (jaredh159) [inline apps/next-evans/components/core/Seo.tsx:12]: hmm... this feels wrong to me, because we're significantly complicating the type passed to this component, just so we can dynamically assemble the og image (as far as i can tell). 

that makes me think that what we really want here is a `ogImage?: string;` prop, which falls back to the feature image when not supplied. then, the doc pages can supply their own og image, instead of all this other data.

File: apps/next-evans/justfile [hunk 3]
```diff
@@ -22,6 +22,9 @@ prisma-generate: pre-prisma && post-prisma
 prisma-db-pull: pre-prisma && post-prisma
   @pnpm prisma db pull --force
 
+test:
+  @pnpm vitest . --watch
```
Comment (jaredh159) [inline apps/next-evans/justfile:26]: 👍

File: apps/next-evans/lib/seo.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { inflect } from './utils';
+
+export function friendPageDescription(
```
Comment (jaredh159) [inline apps/next-evans/lib/seo.ts:3]: hmmm... are you handling spanish books here at all?

or, i guess, maybe at a deeper level, did you decide against just pulling over my `lib/seo.ts` file? all this lang-aware string concat stuff is pretty grungy, and it's all already done (with tests), so it seems like unless we had a compelling reason to rewrite, we should probably use those. sorry if i didn't make it more clear that there was code for this already.

File: apps/next-evans/lib/utils.ts [hunk 5]
```diff
@@ -1,3 +1,7 @@
 export function isNotNullish<T>(x: T | null | undefined): x is T {
   return x !== null && x !== undefined;
 }
+
+export function inflect(str: string, count: number): string {
```
Comment (jaredh159) [inline apps/next-evans/lib/utils.ts:5]: i'm not sure this holds up for spanish too, and i think if we do pull over the other seo file, we won't need this, is my thought.

File: apps/next-evans/pages/audiobooks.tsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import Seo from '@/components/core/Seo';
+
+export const getStaticProps: GetStaticProps<Props> = async () => {
+  const sortedBooks = Object.values(await getAllDocuments())
+    .filter((book) => book.hasAudio)
+    .filter(isNotNullish)
+    .sort((a, b) => {
+      const aDate = a.mostModernEdition.audiobook?.createdAt ?? ``;
+      const bDate = b.mostModernEdition.audiobook?.createdAt ?? ``;
+      return aDate < bDate ? 1 : -1;
```
Comment (jaredh159) [inline apps/next-evans/pages/audiobooks.tsx:26]: 👍

**Review by jaredh159 — CHANGES_REQUESTED** [review]
nice work. left a couple comments, the main gist of which is (unless i'm missing something), i think it would be better to transfer the battle-tested seo code and tests from the prior gatsby version, as they cover spanish, html entities, and some other edge cases.

File: apps/next-evans/components/core/Seo.tsx (hunk 2 above)
Comment (kiahjh) [reply apps/next-evans/components/core/Seo.tsx]: Yeah that makes way more sense

**Comment by jaredh159** [comment]
closing this, i've manually merged in all the stuff from it.  thanks! 👍
